node/client/src/lib.rs

```diff
@@ -178,23 +176,33 @@ impl Client {
             apply_result.authority_proposals,
             shard_block.block_hash(),
         );
-        let authority_mask: Vec<bool> =
-            authorities.iter().map(|a| a.account_id == self.signer.account_id()).collect();
+        // TODO: We should have a proper mask computation once we have a correct consensus.
+        let authority_mask: Vec<bool> = authorities.iter().map(|_| true).collect();
         let signature = shard_block.sign(&self.signer);
         shard_block.add_signature(signature);
         shard_block.authority_mask = authority_mask.clone();
         let signature = block.sign(&self.signer);
         block.add_signature(signature);
         block.authority_mask = authority_mask;
-        self.shard_chain.insert_block(&shard_block.clone());
-        self.beacon_chain.insert_block(block.clone());
-        info!(target: "block_producer", "Block body: {:?}", block.body);
-        info!(target: "block_producer", "Shard block body: {:?}", shard_block.body);
-        io::stdout().flush().expect("Could not flush stdout");
-
-        // Update the authority.
-        self.update_authority(&block.header());
-        (block, shard_block)
+
+        if self.beacon_chain.is_known(&block.hash) {
+            info!(target: "client", "The block was already imported, before we managed to produce it.");
+            io::stdout().flush().expect("Could not flush stdout");
+            None
+        } else {
+            self.shard_chain.insert_block(&shard_block.clone());
+            self.beacon_chain.insert_block(block.clone());
+            info!(target: "client",
+                  "Producing block index: {:?}, beacon = {:?}, shard = {:?}",
+                  block.body.header.index, block.hash, shard_block.hash);
+            io::stdout().flush().expect("Could not flush stdout");
+            // Just produced blocks should be the best in the blockchain.
+            assert_eq!(self.shard_chain.chain.best_block().hash, shard_block.hash);
```

Comment:
this kind of check is awkward in non-testing code

Reply:
Just a regular panic check, like `expect` on `Option`.

Reply:
it just seems weird to arbitrarily define here that we do not trust that `shard_chain` or `beacon_chain` do not work as intended, and rely on assertions and panics, and that everywhere else we are happy with guarantees + tests that the interfaces provide